lib/tutf8e/src/tutf8e.c

```diff
@@ -3,121 +3,200 @@
+
 /* return TUTF8E_INVALID if input character is not convertable */
 /* return TUTF8E_OK for success                                */
+/* return TUTF8E_SAME : ...                                    */
 
-int tutf8e_string_encode(const uint16_t *table, const char *i, char *o, size_t *olen)
+int tutf8e_string_encode(const TUTF8encoder encoder, const char *input, char *o, uint32_t flags)
 {
   int ret;
-  size_t ilen = 0;
+  size_t ilen = strlen(input);
   size_t length = 0;
-  if (!(ret = tutf8e_string_length(table, i, &ilen, &length)))
-  {
-    if (length+1 > *olen) return TUTF8E_TOOLONG;
-    if (!(ret = tutf8e_buffer_encode(table, i, ilen, o, olen)))
-    {
-      o[length] = 0;
-      return TUTF8E_OK;
-    }
+
+  if((ret = tutf8e_string_length(encoder, input, &ilen, &length, flags)) < 0) {
+    return ret;
+  }
+
+  if ((ret = tutf8e_buffer_encode(encoder, input, ilen, o, flags)) < 0) {
+    return ret;
   }
-  return ret;
+
+  o[length] = 0;
+
+  return length;
 }
 
 /* Determine the length of the UTF8 encoding of given input string and table */
 /* return TUTF8E_INVALID if input character is not convertable               */
 /* return TUTF8E_OK for success                                              */
 
-int tutf8e_buffer_length(const uint16_t *table, const char *input, size_t ilen, size_t *length)
+int tutf8e_buffer_length(const TUTF8encoder encoder, const char *input, size_t ilen, size_t *olen, uint32_t flags)
 {
+  const uint16_t *table = (uint16_t *) encoder;
   const unsigned char *i;
+  int changed = 0;
+  int len = 0;
   for (i = (const unsigned char *) input; ilen; ++i, --ilen) {
     const uint16_t c = table[*i];
     if (c<0x80) {
-      ++*length;
+      len++;
+      if(c != *i) changed++;
       continue;
     }
+    changed++;
     if (c<0x800) {
-      *length += 2;
+      len += 2;
       continue;
     }
     if (c<0xffff) {
-      *length += 3;
+      len += 3;
       continue;
     }
-    return TUTF8E_INVALID;
+    switch(flags & TUTF8E_FLAG_INV_MASK) {
+    case TUTF8E_FLAG_INV_KEEP        : len += *i <= 0x80 ? 1 : 2; break;
+    case TUTF8E_FLAG_INV_FAIL        : return  TUTF8E_INVALID;
+    case TUTF8E_FLAG_INV_IGNORE      : break;
+    case TUTF8E_FLAG_INV_REPLACEMENT : len += 3; break;
+    case TUTF8E_FLAG_INV_QUESTION    : len++; break;
+    case TUTF8E_FLAG_INV_COPY        : len++; break;
+    }
   }
-  return TUTF8E_OK;
+  *olen = len;
+  return changed ? TUTF8E_OK : TUTF8E_SAME;
 }
 
 /* UTF8 encode the given input string and table                */
 /* olen input is output buffer size, output is encoded length  */
-/* return TUTF8E_TOOLONG if output buffer insuficient          */
 /* return TUTF8E_INVALID if input character is not convertable */
-/* return TUTF8E_OK for success                                */
+/* return >= 0 size of encoded string                          */
 
-int tutf8e_buffer_encode(const uint16_t *table, const char *input, size_t ilen, char *output, size_t *olen)
```

Comment:
One comment about performance assumptions.

The formulation here with _olen_ it to avoid exceeding the output buffer without pre-computing the length of the encoded output.  Without checking, this change _assumes_ and _requires_ the output buffer to be sufficiently large. 

If we assume there is "ample" space in the output buffer, it should not be necessary to do two steps.  But we do need to check for when we do (occasionally) run out of space.

Reply:
yes.. your original code has has check for output size... that is better. 
But having this dual purpose **\*olen** is al little bit problematic: input=size of buffer, output=size of output. 
Ok could go with that.